scylla/src/transport/query_result.rs

```diff
@@ -0,0 +1,719 @@
+        match self.maybe_first_row::<R>() {
+            Ok(Some(row)) => Ok(row),
+            Ok(None) => Err(FirstRowError::RowsEmpty),
+            Err(MaybeFirstRowError::NotRowsResponse) => Err(FirstRowError::NotRowsResponse),
+            Err(MaybeFirstRowError::TypeCheckFailed(err)) => {
+                Err(FirstRowError::TypeCheckFailed(err))
+            }
+            Err(MaybeFirstRowError::DeserializationFailed(err)) => {
+                Err(FirstRowError::DeserializationFailed(err))
+            }
+        }
+    }
+
+    /// Returns the only received row.\
+    /// Fails if the result is anything else than a single row.
+    pub fn single_row<'res, R: DeserializeRow<'res>>(&'res self) -> Result<R, SingleRowError> {
+        match self.rows::<R>() {
+            Ok(mut rows) => match rows.next() {
+                Some(Ok(row)) => {
+                    if rows.rows_remaining() != 0 {
+                        return Err(SingleRowError::UnexpectedRowCount(
+                            rows.rows_remaining() + 1,
+                        ));
+                    }
+                    Ok(row)
+                }
+                Some(Err(err)) => Err(SingleRowError::DeserializationFailed(err)),
+                None => Err(SingleRowError::UnexpectedRowCount(0)),
+            },
+            Err(RowsError::NotRowsResponse) => Err(SingleRowError::NotRowsResponse),
+            Err(RowsError::TypeCheckFailed(err)) => Err(SingleRowError::TypeCheckFailed(err)),
+        }
+    }
+
+    /// Returns column specifications or None if the result is not of Rows type.
+    #[inline]
+    pub fn column_specs(&self) -> Option<ColumnSpecs> {
+        self.raw_rows
+            .as_ref()
+            .map(|raw_rows| ColumnSpecs::new(&raw_rows.metadata().col_specs))
+    }
+
+    /// Warnings emitted by the database.
+    #[inline]
+    pub fn warnings(&self) -> impl Iterator<Item = &str> {
+        self.warnings.iter().map(String::as_str)
+    }
+
+    /// Paging state of the query that was reached after performing this request.
+    #[inline]
+    pub fn paging_state(&self) -> Option<Bytes> {
+        self.raw_rows.as_ref()?.metadata().paging_state.clone()
+    }
+
+    /// Tracing ID associated with this CQL request.
+    #[inline]
+    pub fn tracing_id(&self) -> Option<Uuid> {
+        self.tracing_id
+    }
+
+    /// Transforms itself into the legacy result type, by eagerly deserializing rows
+    /// into the Row type. This is inefficient, and should only be used during transition
+    /// period to the new API.
+    pub fn into_legacy_result(self) -> Result<LegacyQueryResult, ParseError> {
+        if let Some(raw_rows) = self.raw_rows {
+            let deserialized_rows = raw_rows
+                .rows_iter::<Row>()?
+                .collect::<Result<Vec<_>, DeserializationError>>()?;
+            let serialized_size = raw_rows.rows_size();
+            let metadata = raw_rows.into_metadata();
+            Ok(LegacyQueryResult {
```

Comment:
It can fail in a lot of ways, just because the serialized data can be malformed. In such case, DeserializationError will be returned.
I don't see how this is a complication. Question marks can be added easily.